### PR TITLE
This is a version of Richard's function that uses *maybe* a bit less memory.

### PR DESCRIPTION
**Pull request by lemire**
I am not sure that this is worth it. This is meant to address @kishoreg's concerns.

### CODE REVIEW

**Comment by richardstartin** [comment]
Do you have numbers? The idea was that filling up the slices would avoid most of the work. You don’t do any container intersections unless you know you have a full slice of N containers.

**Comment by richardstartin** [comment]
Let’s benchmark this and take it if it’s better, no need for two versions. I’m not all that enthusiastic about memory improvements at the cost of throughput because the algorithm already avoids a lot of allocation compared to naive_and. If this improves things then it’s progress.

**Comment by lemire** [comment]
> The idea was that filling up the slices would avoid most of the work. You don’t do any container intersections unless you know you have a full slice of N containers.

I think it is exactly the same algorithm. The difference is that you construct a potentially large array where your store the containers, and then reload them from this array.

> Do you have numbers?

We don't have a benchmark right now where your approach uses a lot of memory, except for the allocation of the temporary 1024-longs array. 

For that to happen, you need for numContainers to be large at the same time as bitmaps.length is large. So take one constant but multiply repeated bitmap, containing many containers. So numContainers equals to 4000 and bitmaps.length equals to 100,000. Then your approach will allocate a block of 300 MB each and every time it is called, unlike my approach.

I am not pushing this PR. I just want to illustrate what is possible.

If I wanted to push it, I would design a benchmark to make it look good. But that's not very interesting (creating the benchmark after the fact to make your code look good).

@kishoreg is concerned about memory usage. I'd invite him and other people who have these concerns to have a look at my PR.

**Comment by lemire** [comment]
@richardstartin Let me put it another way. I started answering @kishoreg about design trade-offs and figured it would be easier to show what I meant with code.

We can merge, or not, this function. I have no vested interest. I just wanted to show that we could take your excellent approach and drastically reduce the worst-case memory usage.

I agree that your approach will do fine memory-wise in practice, I expect... But maybe  @kishoreg has other concerns that I cannot foresee.

**Comment by lemire** [comment]
Let me add also that creating a benchmark that shines where the memory usage of your opponent blows up, is something I'd rather not check into RoaringBitmap.

**Comment by lemire** [comment]
We could also create hybrids.

**Comment by richardstartin** [comment]
@lemire I think it's best to merge your changes (i.e. the merging of the last two loops and removal of the 2D containers array) into the existing method, rather than create a new one.

File: RoaringBitmap/src/main/java/org/roaringbitmap/FastAggregation.java [hunk 1]
```diff
@@ -335,6 +335,72 @@ public static RoaringBitmap workShyAnd(RoaringBitmap... bitmaps) {
...
+      }
+      base += 64;
+    }
+    RoaringArray array =
+            new RoaringArray(keys, new Container[numContainers], 0);
+    for (int i = 0; i < numContainers; ++i) {
+      char matching_key = keys[i];
+      Arrays.fill(words, -1L);
+      Container tmp = new BitmapContainer(words, -1);
+      for(RoaringBitmap bitmap: bitmaps) {
```
Comment (richardstartin) [inline RoaringBitmap/src/main/java/org/roaringbitmap/FastAggregation.java:385]: What's the worst case of this loop? Isn't this O(m log n) where m is the number of keys in the intersection and n is the number of containers in the bitmap? 

Can't we stick to camel case?

**Comment by richardstartin** [comment]
To be honest, I don't see a reason not to apply the change to the old method, which will probably be deleted at some point anyway. 

Reusing the buffer is obviously also applicable to the original method. If you want two versions of the method, can you make `workShyAnd` take a buffer as input too, and then provide an overload of `and` which takes a buffer?

So:
```java
RoaringBitmap and(RoaringBitmap... bitmaps) {
  return workShyAnd(new long[1024], bitmaps);
}

RoaringBitmap and(long[] bitmap, RoaringBitmap... bitmaps) {
    return workShyAnd(bitmap, bitmaps);
}

```

**Comment by lemire** [comment]
@richardstartin  I am running a benchmark:

```
$ java -jar jmh/target/benchmarks.jar FastAggregation

Benchmark                                        (constructionStrategy)  (count)  (probability)  (seed)   (size)   Mode  Cnt    Score     Error  Units
FastAggregationRLEStressTest.and                        CONSTANT_MEMORY      100           0.01   99999  1000000  thrpt    5  867.029 ± 170.413  ops/s
FastAggregationRLEStressTest.and                        CONSTANT_MEMORY      100            0.1   99999  1000000  thrpt    5  102.334 ±  17.711  ops/s
FastAggregationRLEStressTest.and                        CONSTANT_MEMORY      100            0.5   99999  1000000  thrpt    5  350.416 ±  36.901  ops/s
FastAggregationRLEStressTest.and                     CONTAINER_APPENDER      100           0.01   99999  1000000  thrpt    5  936.893 ±  90.405  ops/s
FastAggregationRLEStressTest.and                     CONTAINER_APPENDER      100            0.1   99999  1000000  thrpt    5  107.553 ±   3.822  ops/s
FastAggregationRLEStressTest.and                     CONTAINER_APPENDER      100            0.5   99999  1000000  thrpt    5  342.889 ±  27.903  ops/s
FastAggregationRLEStressTest.andBuffer                  CONSTANT_MEMORY      100           0.01   99999  1000000  thrpt    5  920.579 ±  61.322  ops/s
FastAggregationRLEStressTest.andBuffer                  CONSTANT_MEMORY      100            0.1   99999  1000000  thrpt    5   91.479 ±  35.229  ops/s
FastAggregationRLEStressTest.andBuffer                  CONSTANT_MEMORY      100            0.5   99999  1000000  thrpt    5  300.711 ± 105.753  ops/s
FastAggregationRLEStressTest.andBuffer               CONTAINER_APPENDER      100           0.01   99999  1000000  thrpt    5  777.514 ± 424.695  ops/s
FastAggregationRLEStressTest.andBuffer               CONTAINER_APPENDER      100            0.1   99999  1000000  thrpt    5   93.305 ±  11.589  ops/s
FastAggregationRLEStressTest.andBuffer               CONTAINER_APPENDER      100            0.5   99999  1000000  thrpt    5  297.298 ±  26.440  ops/s
FastAggregationRLEStressTest.andBufferMemoryShy         CONSTANT_MEMORY      100           0.01   99999  1000000  thrpt    5  811.430 ± 340.511  ops/s
FastAggregationRLEStressTest.andBufferMemoryShy         CONSTANT_MEMORY      100            0.1   99999  1000000  thrpt    5   84.778 ±  36.852  ops/s
FastAggregationRLEStressTest.andBufferMemoryShy         CONSTANT_MEMORY      100            0.5   99999  1000000  thrpt    5  272.822 ±  32.180  ops/s
FastAggregationRLEStressTest.andBufferMemoryShy      CONTAINER_APPENDER      100           0.01   99999  1000000  thrpt    5  848.879 ± 180.828  ops/s
FastAggregationRLEStressTest.andBufferMemoryShy      CONTAINER_APPENDER      100            0.1   99999  1000000  thrpt    5   94.064 ±  10.410  ops/s
FastAggregationRLEStressTest.andBufferMemoryShy      CONTAINER_APPENDER      100            0.5   99999  1000000  thrpt    5  290.661 ±  29.712  ops/s
FastAggregationRLEStressTest.andMemoryShy               CONSTANT_MEMORY      100           0.01   99999  1000000  thrpt    5  912.764 ± 195.919  ops/s
FastAggregationRLEStressTest.andMemoryShy               CONSTANT_MEMORY      100            0.1   99999  1000000  thrpt    5  104.556 ±  15.336  ops/s
FastAggregationRLEStressTest.andMemoryShy               CONSTANT_MEMORY      100            0.5   99999  1000000  thrpt    5  347.633 ±  70.697  ops/s
FastAggregationRLEStressTest.andMemoryShy            CONTAINER_APPENDER      100           0.01   99999  1000000  thrpt    5  873.314 ± 118.928  ops/s
FastAggregationRLEStressTest.andMemoryShy            CONTAINER_APPENDER      100            0.1   99999  1000000  thrpt    5  106.296 ±   6.539  ops/s
FastAggregationRLEStressTest.andMemoryShy            CONTAINER_APPENDER      100            0.5   99999  1000000  thrpt    5  341.304 ±   8.545  ops/s
```

**Comment by lemire** [comment]
> To be honest, I don't see a reason not to apply the change to the old method, which will probably be deleted at some point anyway.

I can delete your method and override mine as you propose. It is trivial.

(Note that you also have write access to the branch in question. Feel free to hack it directly if you want.)

**Comment by lemire** [comment]
>  I think it's best to merge your changes (i.e. the merging of the last two loops and removal of the 2D containers array) into the existing method, rather than create a new one. 

Please confirm that this is what you want me to do.

**Comment by richardstartin** [comment]
Actually, once you reuse the `words` buffer in the old method, I'm not sure this is the best of the two approaches, this is only better when the intersection itself is very large. So maybe it's better to keep the two separate. I have a change I will apply to your branch.

**Comment by richardstartin** [comment]
This is a space/time tradeoff. With this approach, I think the user may give up up to 5% throughput to avoid allocating an array which will often be tiny, and the allocation is reduced anyway because there are no container transitions. Let's add the new method and users who care about memory can call the one which allocates less.

**Comment by richardstartin** [comment]
E.g. see the benchmark derived from druid. If we're being objective I don't think we should shy away from constructing a benchmark to demonstrate the worst case for any algorithm.

```
Benchmark                                        (constructionStrategy)  (count)  (probability)  (seed)   (size)   Mode  Cnt      Score      Error  Units
FastAggregationRLEStressTest.and                     CONTAINER_APPENDER       10           0.01   99999  1000000  thrpt    5   4855.770 ▒  373.631  ops/s
FastAggregationRLEStressTest.and                     CONTAINER_APPENDER       10            0.1   99999  1000000  thrpt    5  11604.443 ▒   40.282  ops/s
FastAggregationRLEStressTest.and                     CONTAINER_APPENDER       10            0.5   99999  1000000  thrpt    5   6752.779 ▒  120.192  ops/s
FastAggregationRLEStressTest.and                     CONTAINER_APPENDER      100           0.01   99999  1000000  thrpt    5    639.729 ▒   22.134  ops/s
FastAggregationRLEStressTest.and                     CONTAINER_APPENDER      100            0.1   99999  1000000  thrpt    5     66.867 ▒    1.848  ops/s
FastAggregationRLEStressTest.and                     CONTAINER_APPENDER      100            0.5   99999  1000000  thrpt    5    218.860 ▒   10.424  ops/s
FastAggregationRLEStressTest.andMemoryShy            CONTAINER_APPENDER       10           0.01   99999  1000000  thrpt    5   4721.572 ▒  269.369  ops/s
FastAggregationRLEStressTest.andMemoryShy            CONTAINER_APPENDER       10            0.1   99999  1000000  thrpt    5  10329.444 ▒ 2976.924  ops/s
FastAggregationRLEStressTest.andMemoryShy            CONTAINER_APPENDER       10            0.5   99999  1000000  thrpt    5   6608.043 ▒  229.798  ops/s
FastAggregationRLEStressTest.andMemoryShy            CONTAINER_APPENDER      100           0.01   99999  1000000  thrpt    5    567.643 ▒   96.512  ops/s
FastAggregationRLEStressTest.andMemoryShy            CONTAINER_APPENDER      100            0.1   99999  1000000  thrpt    5     66.356 ▒    0.860  ops/s
FastAggregationRLEStressTest.andMemoryShy            CONTAINER_APPENDER      100            0.5   99999  1000000  thrpt    5    137.922 ▒    4.764  ops/s
FastAggregationRLEStressTest.andBuffer               CONTAINER_APPENDER       10           0.01   99999  1000000  thrpt    5   5132.889 ▒  264.818  ops/s
FastAggregationRLEStressTest.andBuffer               CONTAINER_APPENDER       10            0.1   99999  1000000  thrpt    5  11980.194 ▒  156.344  ops/s
FastAggregationRLEStressTest.andBuffer               CONTAINER_APPENDER       10            0.5   99999  1000000  thrpt    5   6186.253 ▒   51.083  ops/s
FastAggregationRLEStressTest.andBuffer               CONTAINER_APPENDER      100           0.01   99999  1000000  thrpt    5    613.217 ▒   25.316  ops/s
FastAggregationRLEStressTest.andBuffer               CONTAINER_APPENDER      100            0.1   99999  1000000  thrpt    5     60.029 ▒    1.385  ops/s
FastAggregationRLEStressTest.andBuffer               CONTAINER_APPENDER      100            0.5   99999  1000000  thrpt    5    222.801 ▒    1.932  ops/s
FastAggregationRLEStressTest.andBufferMemoryShy      CONTAINER_APPENDER       10           0.01   99999  1000000  thrpt    5   5301.013 ▒  290.699  ops/s
FastAggregationRLEStressTest.andBufferMemoryShy      CONTAINER_APPENDER       10            0.1   99999  1000000  thrpt    5  10887.899 ▒   73.470  ops/s
FastAggregationRLEStressTest.andBufferMemoryShy      CONTAINER_APPENDER       10            0.5   99999  1000000  thrpt    5   6274.564 ▒   40.254  ops/s
FastAggregationRLEStressTest.andBufferMemoryShy      CONTAINER_APPENDER      100           0.01   99999  1000000  thrpt    5    603.944 ▒   11.274  ops/s
FastAggregationRLEStressTest.andBufferMemoryShy      CONTAINER_APPENDER      100            0.1   99999  1000000  thrpt    5     59.613 ▒    0.230  ops/s
FastAggregationRLEStressTest.andBufferMemoryShy      CONTAINER_APPENDER      100            0.5   99999  1000000  thrpt    5    219.355 ▒   13.373  ops/s
```

File: RoaringBitmap/src/main/java/org/roaringbitmap/FastAggregation.java [hunk 2]
```diff
@@ -37,7 +37,30 @@ public static RoaringBitmap and(Iterator<? extends RoaringBitmap> bitmaps) {
...
-      return workShyAnd(bitmaps);
+      return workShyAnd(new long[1024], bitmaps);
+    }
+    return naive_and(bitmaps);
+  }
+
+  /**
+   * Compute the AND aggregate.
+   *
+   * In practice, calls {#link naive_and}
```
Comment (lemire) [inline RoaringBitmap/src/main/java/org/roaringbitmap/FastAggregation.java:48]: I think that this comment is not true.

File: RoaringBitmap/src/main/java/org/roaringbitmap/buffer/BufferFastAggregation.java [hunk 3]
```diff
@@ -27,7 +27,32 @@
...
-      return workShyAnd(bitmaps);
+      return workShyAnd(new long[1024], bitmaps);
+    }
+    return naive_and(bitmaps);
+  }
+
+  /**
+   * Compute the AND aggregate.
+   *
+   * In practice, calls {#link naive_and}
```
Comment (lemire) [inline RoaringBitmap/src/main/java/org/roaringbitmap/buffer/BufferFastAggregation.java:38]: This comment is probably not true.

**Comment by lemire** [comment]
> This is a space/time tradeoff. With this approach, I think the user may give up up to 5% throughput to avoid allocating an array which will often be tiny, and the allocation is reduced anyway because there are no container transitions. Let's add the new method and users who care about memory can call the one which allocates less.

This sounds good.

Can you have a look at the comments that I flagged. I think that you introduced two lies. :-)

**Comment by richardstartin** [comment]
resolved

**Comment by lemire** [comment]
@richardstartin If you are happy, I will merge.

**Comment by richardstartin** [comment]
Looking at allocation rate, this doesn't really make much difference on the cases we're measuring

```

Benchmark                                                                     (constructionStrategy)  (count)  (probability)  (seed)   (size)   Mode  Cnt       Score         Error   Units
FastAggregationRLEStressTest.and:▒gc.alloc.rate                                   CONTAINER_APPENDER       10           0.01   99999  1000000  thrpt    5      68.985 ▒       3.039  MB/sec
FastAggregationRLEStressTest.and:▒gc.alloc.rate                                   CONTAINER_APPENDER       10            0.1   99999  1000000  thrpt    5     945.149 ▒      73.921  MB/sec
FastAggregationRLEStressTest.and:▒gc.alloc.rate                                   CONTAINER_APPENDER       10            0.5   99999  1000000  thrpt    5     568.803 ▒      33.396  MB/sec
FastAggregationRLEStressTest.and:▒gc.alloc.rate                                   CONTAINER_APPENDER      100           0.01   99999  1000000  thrpt    5      10.904 ▒       0.366  MB/sec
FastAggregationRLEStressTest.and:▒gc.alloc.rate                                   CONTAINER_APPENDER      100            0.1   99999  1000000  thrpt    5       5.742 ▒       0.101  MB/sec
FastAggregationRLEStressTest.and:▒gc.alloc.rate                                   CONTAINER_APPENDER      100            0.5   99999  1000000  thrpt    5      19.524 ▒       0.306  MB/sec
FastAggregationRLEStressTest.andMemoryShy:▒gc.alloc.rate                          CONTAINER_APPENDER       10           0.01   99999  1000000  thrpt    5      61.712 ▒       7.650  MB/sec
FastAggregationRLEStressTest.andMemoryShy:▒gc.alloc.rate                          CONTAINER_APPENDER       10            0.1   99999  1000000  thrpt    5     943.880 ▒      30.908  MB/sec
FastAggregationRLEStressTest.andMemoryShy:▒gc.alloc.rate                          CONTAINER_APPENDER       10            0.5   99999  1000000  thrpt    5     571.208 ▒       8.122  MB/sec
FastAggregationRLEStressTest.andMemoryShy:▒gc.alloc.rate                          CONTAINER_APPENDER      100           0.01   99999  1000000  thrpt    5       8.051 ▒       1.129  MB/sec
FastAggregationRLEStressTest.andMemoryShy:▒gc.alloc.rate                          CONTAINER_APPENDER      100            0.1   99999  1000000  thrpt    5       5.443 ▒       0.176  MB/sec
FastAggregationRLEStressTest.andMemoryShy:▒gc.alloc.rate                          CONTAINER_APPENDER      100            0.5   99999  1000000  thrpt    5      18.445 ▒       0.580  MB/sec
FastAggregationRLEStressTest.andBuffer:▒gc.alloc.rate                             CONTAINER_APPENDER       10           0.01   99999  1000000  thrpt    5      75.860 ▒       2.154  MB/sec
FastAggregationRLEStressTest.andBuffer:▒gc.alloc.rate                             CONTAINER_APPENDER       10            0.1   99999  1000000  thrpt    5     966.632 ▒      21.480  MB/sec
FastAggregationRLEStressTest.andBuffer:▒gc.alloc.rate                             CONTAINER_APPENDER       10            0.5   99999  1000000  thrpt    5     537.416 ▒       9.896  MB/sec
FastAggregationRLEStressTest.andBuffer:▒gc.alloc.rate                             CONTAINER_APPENDER      100           0.01   99999  1000000  thrpt    5      11.204 ▒       0.197  MB/sec
FastAggregationRLEStressTest.andBuffer:▒gc.alloc.rate                             CONTAINER_APPENDER      100            0.1   99999  1000000  thrpt    5       5.209 ▒       0.116  MB/sec
FastAggregationRLEStressTest.andBuffer:▒gc.alloc.rate                             CONTAINER_APPENDER      100            0.5   99999  1000000  thrpt    5      19.745 ▒       1.215  MB/sec
FastAggregationRLEStressTest.andBufferMemoryShy:▒gc.alloc.rate                    CONTAINER_APPENDER       10           0.01   99999  1000000  thrpt    5      78.094 ▒       2.064  MB/sec
FastAggregationRLEStressTest.andBufferMemoryShy:▒gc.alloc.rate                    CONTAINER_APPENDER       10            0.1   99999  1000000  thrpt    5     872.449 ▒      61.661  MB/sec
FastAggregationRLEStressTest.andBufferMemoryShy:▒gc.alloc.rate                    CONTAINER_APPENDER       10            0.5   99999  1000000  thrpt    5     361.426 ▒     206.814  MB/sec
FastAggregationRLEStressTest.andBufferMemoryShy:▒gc.alloc.rate                    CONTAINER_APPENDER      100           0.01   99999  1000000  thrpt    5       8.075 ▒       0.353  MB/sec
FastAggregationRLEStressTest.andBufferMemoryShy:▒gc.alloc.rate                    CONTAINER_APPENDER      100            0.1   99999  1000000  thrpt    5       4.849 ▒       0.280  MB/sec
FastAggregationRLEStressTest.andBufferMemoryShy:▒gc.alloc.rate                    CONTAINER_APPENDER      100            0.5   99999  1000000  thrpt    5      18.247 ▒       0.916  MB/sec
```

**Comment by lemire** [comment]
I did not expect a difference in our benchmark.

**Comment by kishoreg** [comment]
can you point me to the benchmark code

**Comment by richardstartin** [comment]
@kishoreg if you’re trying to use this in Pinot and it’s not working well it would be interesting to have some benchmark data

**Pull request merged**